### ISSUE TITLE
Add option to allow saving of reserved filenames on Windows

### ISSUE THREAD

**Issue by rasa** [syncthing/syncthing#9623]
### Feature description

Add an advanced option to allow saving the reserved filenames:

```
"CON", "PRN", "AUX", "NUL",
"COM0", "COM1", "COM2", "COM3", "COM4",
"COM5", "COM6", "COM7", "COM8", "COM9",
"LPT0", "LPT1", "LPT2", "LPT3", "LPT4",
"LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
"CONIN$", "CONOUT$" // disallowed by Windows, but not currently rejected by Syncthing
```
with or without a filename extension, and filenames that end with a period (.) or space (\x20) character.

### Problem or use case

Syncthing currently [disallows](https://github.com/syncthing/syncthing/blob/de0b4270df8dd65d5a1992408191495d1f01dd00/lib/model/folder_sendrecv.go#L351) these filenames on Windows, but if you remove this check, Windows will save these filenames just fine, as we add the `\\?\` prefix to all filenames [here](https://github.com/syncthing/syncthing/blob/de0b4270df8dd65d5a1992408191495d1f01dd00/lib/fs/basicfs.go#L371).

We added this check to protect users from creating these filenames, as users cannot delete or rename these filenames using Windows Explorer, or cmd.exe. But experienced users can create and manipulate these files using many different methods, such as WSL, GitBash, and some Windows Explorer replacement apps.

This proposal suggests adding an option on a folder's advanced options page, that when set to `true`, Syncthing would simply skips the above check. I would suggest the field be named "Allow Reserved Filenames", and always default to `false`.

### Alternatives or workarounds

Replace Windows and Android devices with Linux and iOS devices. :)

**Comment by calmh** [comment]
Y tho, what's the use case?

**Comment by rasa** [comment]
@calmh I added

> As a user, I wish to avoid Syncthing reporting that a folder is out of sync, due to the filename being incompatible with Windows Explorer.

to the above text. Is that acceptable?

**Comment by calmh** [comment]
No, why do you want to specifically sync a file called COM8 or whatever to your windows computer? Under which circumstances is this a common enough thing people want to do?

**Comment by rasa** [comment]
> No, why do you want to specifically sync a file called COM8 or whatever to your windows computer? Under which circumstances is this a common enough thing people want to do?

Agreed, COM8 is pretty uncommon, but GitForWindows allows users to checkout repos containing "reserved" filenames. Here's a list of those on GitHub:

1. *.: [128k](https://github.com/search?q=+path%3A%2F%5C.%24%2F&type=code)
2. con.*: [235](https://github.com/search?q=+path%3A%2F%5Econ%5C.%2F&type=code)
3. aux.*: [100](https://github.com/search?q=+path%3A%2F%5Eaux%5C.%2F&type=code)
4. *\s: [100](https://github.com/search?q=+path%3A%2F+%24%2F&type=code)
5. nul: [49](https://github.com/search?q=+path%3A%2F%5Enul%24%2F&type=code)
6. aux: [20](https://github.com/search?q=+path%3A%2F%5Eaux%24%2F&type=code)
7. con: [10](https://github.com/search?q=+path%3A%2F%5Econ%24%2F&type=code)

**Comment by calmh** [comment]
Yeah... so there's over a billion files in repos on GitHub, so essentially what I'm hearing is that this doesn't happen to any degree that's above the noise level. I don't want to add config knobs for crazy rare situations, really.

**Comment by rasa** [comment]
@calmh Can you define "crazy rare"? Is that one-in-a-million? Note that 128,000 is the number of public **_repos_**, not files.

And Git for Windows, and many other programs support creating these files, so I don't really understand the reticence, given that it's a option that defaults to off.

**Comment by calmh** [comment]
So, crazy was a bad word choice for many reasons. Let's disregard that.

Every option has a cost in complexity, documentation, testing, support over time, etc.  If it's not likely to be ever used by even 1% of users, I don't think there's a good ROI. This, to me, seems like something in that niche.
